url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/math/Generic_BigDecimal.java

```diff
@@ -0,0 +1,7278 @@
+        if (x.compareTo(BigDecimal.ZERO) == 0) {
+            return BigDecimal.ZERO;
+        }
+        if (y.compareTo(BigDecimal.ZERO) == 0) {
+            throw new ArithmeticException("Attempted division by zero...");
+        }
+        if (y.compareTo(BigDecimal.ONE) == 0) {
+            return new BigDecimal(x.toString());
+        }
+        if (x.compareTo(y) == 0) {
+            return BigDecimal.ONE;
+        }
+        return divideNoCaseCheckNoRounding(
+                x,
+                y);
+    }
+
+    /**
+     * @param x
+     * @param y
+     * @param decimalPlaces
+     * @param a_RoundingMode
+     * @return x/y if it can be accurately represented as a BigDecimal. Throws a
+     * ArithmeticException otherwise
+     */
+    private static BigDecimal divideNoCaseCheckNoRounding(
+            BigDecimal x,
+            BigDecimal y) {
+        BigDecimal result = x.divide(y, RoundingMode.UNNECESSARY);
+        return result;
+    }
+
+    /**
+     * Creates and returns a new list containing all the numbers in list divided
+     * by divisor using decimalPlaces and roundingMode and rounding if
+     * necessary.
+     * @param roundingMode
+     * @param decimalPlaces
+     * @return 
+     */
+    public static ArrayList<BigDecimal> divideRoundIfNecessary(
+            ArrayList<BigDecimal> list,
+            BigDecimal divisor,
+            int decimalPlaces,
+            RoundingMode roundingMode) {
+        ArrayList<BigDecimal> result = new ArrayList<BigDecimal>();
+        Iterator<BigDecimal> ite = list.iterator();
+        while (ite.hasNext()) {
+            BigDecimal value = ite.next();
+            BigDecimal dividedValue = Generic_BigDecimal.divideRoundIfNecessary(
+                    value,
+                    divisor,
+                    decimalPlaces,
+                    roundingMode);
+            result.add(dividedValue);
+        }
+        return result;
+    }
+
+    /**
+     * @param x
+     * @param y
+     * @param decimalPlaces
+     * @param a_RoundingMode
+     * @return x/y then rounded;
+     */
+    public static BigDecimal divideRoundIfNecessary(
+            BigDecimal x,
+            BigInteger y,
+            int decimalPlaces,
+            RoundingMode a_RoundingMode) {
+        // Deal with special cases
+        if (x.compareTo(BigDecimal.ZERO) == 0) {
+            return BigDecimal.ZERO;
+        }
+        if (y.compareTo(BigInteger.ZERO) == 0) {
+            throw new ArithmeticException("Attempted division by zero...");
+        }
+        if (y.compareTo(BigInteger.ONE) == 0) {
+            return new BigDecimal(x.toString());
+        }
+        BigDecimal y_BigDecimal = new BigDecimal(y);
+        if (x.compareTo(y_BigDecimal) == 0) {
+            return BigDecimal.ONE;
+        }
+        return divideNoCaseCheckRoundIfNecessary(
+                x,
+                y_BigDecimal,
+                decimalPlaces,
+                a_RoundingMode);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 834:851, 922:939.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.